File: functions/alsa-lib/variants/c/0_alaw_to_s16_3.c
#include <stdint.h>

static int alaw_to_s16(unsigned char a_val)
{
	int t = (a_val ^ 0x55) & 0x7F;
	int seg = (t & 0x70) >> 4;
	t = (t & 0x0F | 0x10) << seg;
	t = seg ? (t << 3) + 0x100 : t << 4;
	return (a_val & 0x80) != 0 ? t : -t;
}


int main(){
    alaw_to_s16('a');
}

File: functions/alsa-lib/variants/go/0_alaw_to_s16_3.go
package main

func alawToS16(aVal byte) int32 {
	aVal ^= 0x55
	t := int32(aVal&0x7f) << 3
	if t < 128 {
		t += 8
	} else {
		t += 0x800 + ((t & 0x70) << 4)
		t <<= ((t >> 4) & 0x07) - 1
	}
	if aVal&0x80 != 0 {
		return t
	}
	return -t
}


func main(){
    alawToS16('a')
}
